src/Fantomas/TokenParser.fs

```diff
@@ -90,91 +90,189 @@ let private createHashToken lineNumber content fullMatchedLength offset =
             CharClass = FSharpTokenCharKind.WhiteSpace
             FSharpTokenTriggerClass = FSharpTokenTriggerClass.None
             Tag = 0
-            FullMatchedLength = fullMatchedLength } }
+            FullMatchedLength = String.length content } }
 
-type private SourceCodeState =
+type SourceCodeState =
     | Normal
-    | InsideSingleQuoteString
-    | InsideTripleQuoteString
+    | InsideString
+    | InsideTripleQuoteString of int
+    | InsideMultilineComment
 
-let rec private getTokenizedHashes sourceCode =
-    let processLine content (trimmed: string) lineNumber fullMatchedLength offset =
-        let contentLength = String.length content
+let rec private getTokenizedHashes (sourceCode: string): Token list =
+    let hasNoHashDirectiveStart (source: string) = not (source.Contains("#if"))
 
-        let tokens =
-            tokenize [] (trimmed.Substring(contentLength))
+    if hasNoHashDirectiveStart sourceCode then
+        []
+    else
+        let doubleQuoteChar = '"'
+        let backSlashChar = '\\'
+        let openParenChar = '('
+        let closeParenChar = ')'
+        let asteriskChar = '*'
+        let newlineChar = '\n'
+        let hashChar = '#'
+        let spaceChar = ' '
+        let isDoubleQuoteChar = (=) doubleQuoteChar
+        let isBackSlashChar = (=) backSlashChar
+        let isNoBackSlashChar v = v <> backSlashChar
+
+        let tripleQuotes =
+            (doubleQuoteChar, doubleQuoteChar, doubleQuoteChar)
+
+        let isOpenParenChar = (=) openParenChar
```

Comment:
Idle thought: could this all be represented better with active patterns? Instead of all these checks, you could have them all play nicely in a `match` statement.

Reply:
> Idle thought

You mean genius advice! The code to me is indeed more readable with the active patterns.